edb/pgsql/delta_ext_ai.py

```diff
@@ -0,0 +1,678 @@
+        index_name = common.get_index_backend_name(
+            index.id, module_name, catenate=False, aspect=f'{d}_index'
+        )
+
+        pg_index = dbops.Index(
+            name=index_name[1],
+            table_name=table_name,
+            exprs=[f"__ext_ai_{idx_id}_embedding_{d}__"],
+            with_clause=with_clause,
+            unique=False,
+            predicate=predicate_src,
+            metadata={
+                'schemaname': str(index.get_name(schema)),
+                'kwargs': sql_kwarg_exprs,
+                'code': f'hnsw (__col__ {opclass})',
+                'dimensions': str(d),
+                'distance_function': str(df),
+            },
+        )
+        ops.add_command(dbops.CreateIndex(pg_index))
+
+    index_expr = index.get_expr(schema)
+    assert index_expr is not None
+    dep_cols = []
+    assert index_expr.refs is not None
+    for obj in index_expr.refs.objects(schema):
+        if isinstance(obj, s_props.Property):
+            ptrinfo = types.get_pointer_storage_info(obj, schema=schema)
+            dep_cols.append(ptrinfo.column_name)
+
+    ops.add_command(
+        _pg_create_trigger(schema, index, table_name, dep_cols))
+    ops.add_command(
+        _pg_create_ai_embeddings_source_view(index, options, schema))
+
+    return ops
+
+
+def _pg_delete_ai_embeddings(
+    index: s_indexes.Index,
+    drop_index: dbops.Command,
+    schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    table_name = common.get_index_table_backend_name(index, schema)
+    idx_id = index.id.hex
+
+    ops = dbops.CommandGroup()
+    ops.add_command(drop_index)
+
+    ops.add_command(_pg_drop_trigger(index, table_name))
+    ops.add_command(_pg_drop_ai_embeddings_source_view(index, schema))
+
+    # When the ObjectType is being deleted, we don't drop the index,
+    # as it will get dropped with the parent table.
+    # The same goes for the __ext_ai_embedding{N}__ columns.
+    source_drop = isinstance(drop_index, dbops.NoOpCommand)
+    if not source_drop:
+        table_name = common.get_index_table_backend_name(index, schema)
+
+        dimensions = index.must_get_json_annotation(
+            schema,
+            sn.QualName("ext::ai", "embedding_model_output_dimensions"),
+            checked.CheckedList[int],
+        )
+
+        alter_table = dbops.AlterTable(table_name)
+
+        for d in dimensions:
+            alter_table.add_operation(
+                dbops.AlterTableDropColumn(
+                    dbops.Column(
+                        name=f'__ext_ai_{idx_id}_embedding_{d}__',
+                        type=('edgedb', f'vector({d})'),
+                    )
+                )
+            )
+
+        ops.add_command(alter_table)
+
+    return ops
+
+
+def _pg_create_trigger(
+    schema: s_schema.Schema,
+    index: s_indexes.Index,
+    table_name: tuple[str, str],
+    dep_cols: list[str],
+) -> dbops.Command:
+    ops = dbops.CommandGroup()
```

Comment:
Mention what the trigger does?

Reply:
And maybe add a comment that we might want to share some machinery with FTS eventually